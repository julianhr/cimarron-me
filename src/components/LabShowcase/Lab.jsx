import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { css } from '@emotion/core'

import ImageLoader from '../library/ImageLoader'
import AnchorTag from '../library/AnchorTag'


const baseArticle = css`
  border-radius: 5px;
  display: flex;
  flex-direction: column;
`

const STYLES = {
  link: css`
    ${baseArticle}
    max-width: 250px;
    text-decoration: none;
    color: unset;
    background: #ededed;
    padding: 5px 12px 12px;
    will-change: box-shadow;
    transition: box-shadow 30ms ease-in;

    :hover {
      box-shadow: 1px 1px 4px #bbb;
    }
  `
}

const Article = styled.article`
  ${baseArticle}
  padding: 10px 15px;
  width: 100%;
  height: 220px;
  align-items: center;
`

const TechRoot = styled.section`
  display: flex;
  flex-wrap: wrap;
  font-size: 13px;
`

const Tech = styled.div`
  margin: 2px;
  padding: 2px 7px;
`

function Lab({ title, stack, urlPath, thumbnailSrc, isLinkRouted }) {
  const getTechStyle = (index) => {
    const hue = (330 + index * 30) % 360
    const sat = 60
    const lum = 85
    return { background: `hsl(${hue}, ${sat}%, ${lum}%)` }
  }

  const renderTech = stack => (
    stack.map((tech, i) => (
      <Tech
        key={i}
        css={getTechStyle(i)}
      >
        {tech}
      </Tech>
    ))
  )

  return (
    <AnchorTag
      url={urlPath}
      rootStyle={STYLES.link}
      isLinkRouted={isLinkRouted}
    >
      <Article>
        <ImageLoader
          imgSrc={thumbnailSrc}
          maxWidth={200}
          maxHeight={120}
          styles={{ root: { marginBottom: 20 } }}
        />
        <h3>{title}</h3>
      </Article>
      <TechRoot>
        {renderTech(stack)}
      </TechRoot>
    </AnchorTag>
  )
}

Lab.propTypes = {
  thumbnailSrc: PropTypes.string.isRequired,
  stack: PropTypes.arrayOf(PropTypes.string).isRequired,
  title: PropTypes.string.isRequired,
  urlPath: PropTypes.string.isRequired,
  isLinkRouted: PropTypes.bool.isRequired,
}

export default Lab
