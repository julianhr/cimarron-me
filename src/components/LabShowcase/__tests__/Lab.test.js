import React from 'react'
import { render } from 'enzyme'

import Lab from '../Lab'
import imgSrc from '~/__tests__/__fixtures__/imgSrc'
import MockApp from '~/__tests__/__mocks__/MockApp'


const testProps = {
  thumbnailSrc: imgSrc,
  stack: 'React Node.js'.split(' '),
  title: 'Test Title',
  urlPath: 'https://test.com',
  isLinkRouted: false,
}

const renderApp = props => render(<MockApp><Lab {...props} /></MockApp>)

describe('Lab', () => {
  it('should match snapshot with all props', () => {
    expect(renderApp(testProps)).toMatchSnapshot()
  })

  describe('props', () => {
    it('thumbnailSrc is required', () => {
      const props = {...testProps}
      delete props.thumbnailSrc
      expect(() => renderApp(props)).toThrow()
    })

    it('stack is required', () => {
      const props = {...testProps}
      delete props.stack
      expect(() => renderApp(props)).toThrow()
    })

    it('title is required', () => {
      const props = {...testProps}
      delete props.title
      expect(() => renderApp(props)).toThrow()
    })

    it('urlPath is required', () => {
      const props = {...testProps}
      delete props.urlPath
      expect(() => renderApp(props)).toThrow()
    })

    it('isLinkRouted is required', () => {
      const props = {...testProps}
      delete props.isLinkRouted
      expect(() => renderApp(props)).toThrow()
    })
  })
})
