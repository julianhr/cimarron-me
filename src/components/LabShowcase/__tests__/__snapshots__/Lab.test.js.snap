// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lab should match snapshot with all props 1`] = `
<a
  class="css-4yofyo-baseArticle-link-link"
  href="https://test.com"
  rel="noopener noreferrer"
>
  <article
    class="css-1dt8rsn-Article-baseArticle e1lgusa60"
  >
    <figure
      class="css-1t2crv9-Root-ImageLoader eeswxza0"
    >
      <img
        class="css-1k90mm3"
        src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
      />
    </figure>
    <h3
      class="css-1brvuev-H3 e1lgusa63"
    >
      Test Title
    </h3>
  </article>
  <section
    class="css-1k4ntzt-TechRoot e1lgusa61"
  >
    <div
      class="css-if30yk-Tech e1lgusa62"
    >
      React
    </div>
    <div
      class="css-1h4zuz1-Tech e1lgusa62"
    >
      Node.js
    </div>
  </section>
</a>
`;
